### ISSUE TITLE
Elusive refc binary memory leaks in index_updater and replicator_worker processes

### ISSUE THREAD

**Issue by AlexanderKaraberov** [apache/couchdb#2232]
## Description

After we had switched to sharding all our databases by default with a sharding factor of 8 (`q=8`) I started to observe an increased RAM consumption by a CouchDB node. This is totally explainable because there are now much more `couch_db_updater`, `index_updater`, `couch_index` and `couch_file` processes each of those consuming some non-zero amount of memory and we did account for this memory increase. But this growth doesn't seem to be proportional to what we have predicted and Linux kernel reported RSS for `beam.smp` jumped from 6-7GB to 22GB in some cases. 
I was suspecting `refc` binaries are to blame here therefore I ran a function which takes a snapshot of the number of binary refs in all VM processes, then GCs the node entirely, and then takes another snapshot and makes a diff (Fred Herbert's `recon` library which we use for our CouchDB nicely abstracts this). I've managed to free almost 10GB on the node. After this I waited for ~15-20 mins. All of this was happening during the busy day on one of our main servers and `msacc` reported that node was under heavy utilisation. So after 20 mins `RSS` for `beam.smp` process went up again by 3.2GB and `VSZ` by 3.56GB. Suffice it to say that after running the same routine memory reports went back to previous values. By looking at the `bin_leak` report  of how many individual binaries were held and then freed by each process as a delta I've noticed that top most were `couch_replicator_worker`,   `couch_db_updater` and `couch_index_updater`. I think this partially explains current situation because those are definitely not short-lived processes and therefore hold ProcBins to big binary content of documents, etc. 
To be honest I don't see a quick fix here except for some major refactoring and introducing short-lived processes or some temporary one-off processes which eventually might not be compatible with current indexer and replicator architecture. But perhaps there are other approaches which we may incorporate in terms of CouchDB utilisation? One of them is switching to using cluster with syncs over Erlang distribution instead of a heavy replication workflow. This will definitely eliminate `couch_replicator_worker`'s contribution to these leaks and we are currently in the process of this migration. Maybe there are more places to fix?

## Expected Behaviour

In the ideal world I would like not to see these elusive microscopic leaks which in the grand scheme of things (sharding with 8/16 + 800 heavily utilised DBs + replications) result in 10-12GB of RAM taken for nothing. 
`ERL_FULLSWEEP_AFTER 0 or some_small_value` is not an option unfortunately because it slows down node performance and by looking at the microstate accounting I was not happy to see some schedulers at 100% GC. For now we decided to adhere to do a fullsweep garbage collection of all node processes manually at one hour intervals from one of our auxiliary Erlang VM monitoring processes connected to CouchDB as `hidden`. This seems to work well without the downsides of `ERL_FULLSWEEP_AFTER`. But still this is a kludge and doesn't sound like a good approach. 

## Your Environment

* CouchDB Version used: 2.3.0 + custom patch set
* Operating System and version: Debian 9 (stretch), Linux kernel 4.9.189.
* OTP version: 21.3.8.6

**Comment by davisp** [comment]
The couch_replicator_worker is probably straightforward to fix. We could just add a hibernate after a batch is flushed [1] to trigger a thorough GC. There was a similar GC optimization done quite a long time ago for this same issue.

For the couch_index_updater, is that the gen_server process, or the anonymous worker process actually performing the updates?

If it were the gen_server, then likely adding a similar hibernate option after every update would likely be the best approach. I'm somewhat uncertain what'd be accumulating binaries there. Perhaps a largish reduce value? Though it does have all the usual hallmarks of a process that passes binaries around but doesn't do many reductions of its own.

If its not the gen_server then we'll have to do a bit more investigating because that process already invokes garbage collection after every document processed. So having it getting into a state where there's lots of accumulated binaries would be fairly unexpected. Not to manage that its already doing quite a few reductions which should also trigger GC fairly quickly. Certainly I'd expect quickly enough that a GC-the-world would release anything meaningful on a regular basis.

[1] https://github.com/apache/couchdb/blob/2.3.1/src/couch_replicator/src/couch_replicator_worker.erl#L174

**Comment by AlexanderKaraberov** [comment]
Hi @davisp ,
Thank you for stepping in. Indeed, fix for `couch_replicator_worker` seems to a nice and quick fix especially considering that those processes always hold the biggest number of binary refs as well as always are the top most for `process_info(binary)`. I will definitely try this fix.

Regarding other smaller leaks I was actually not very precise because there are three types of processes which consume less space but due to a sheer amount of them amplified by sharding, they contribute a lot to a total RAM consumed by `beam.smp`. These are notably (sorted by number of binaries freed):

1. `couch_db_updater`
2. `couch_index`
3. `couch_index_updater`

I've repeated my tests several times on various production nodes to exclude variance but distribution is almost the same. When I inspect binaries of the mentioned processes I see a lot of repeated `BinaryId`s:
```erlang
{binary,[{139849336815280,30512,304},
         {139849336815280,30512,304},
         {139849336815280,30512,304},
         {139849336815280,30512,304},
         {139849336815280,30512,304},
```
This pattern is very consistent and doesn't look good for me. Typical amount of binaries when I sort by `length(info(binary))` is around 2300-3000. Unfortunately at this moment we are not running a build of BEAM VM with debug symbols therefore I can't use neither `gdb` to print raw content pointed by those `BinaryId`s, nor use `ETP` commands in order to look deeper and actually understand what are those binaries. But perhaps there are some NIFs which I can use leverage for this? (one thing is clear their size is always more than 64 bytes hence we have off-heap ones here).

> For the couch_index_updater, is that the gen_server process, or the anonymous worker process actually performing the updates?

Hm, it looks like it's a worker process [according to a code here](https://github.com/apache/couchdb/blob/2.3.1/src/couch_index/src/couch_index_updater.erl#L71).  But calling info for the process which reclaimed the most binaries I see this:
```
{'$initial_call',{couch_index_updater,init,1}}]},
```
that's how I deduced it is `couch_index_updater`. Same for `couch_db_updater`.

> because that process already invokes garbage collection after every document processed.

Yes you [are right](https://github.com/apache/couchdb/blob/2.3.1/src/couch_index/src/couch_index_updater.erl#L182) but perhaps there more places here which need this?
